integration/tests/cook/test_basic.py

```diff
@@ -941,3 +942,119 @@ def check_unscheduled_reason():
+                unique_reason = [r for r in unscheduled_jobs['reasons'] if r['reason'] ==
+                                 "The job couldn't be placed on any available hosts."][0]
+                self.assertEqual("unique-host-placement-group-constraint",
+                                 unique_reason['data']['reasons'][0]['reason'],
+                                 unique_reason)
+        finally:
+            util.kill_jobs(self.cook_url, [uuids])
+
+    def test_balanced_host_constraint(self):
+        state = util.get_mesos_state(self.mesos_url)
+        num_hosts = len(state['slaves'])
+        minimum_hosts = min(num_hosts, 10)
+        group = {'uuid': str(uuid.uuid4()),
+                 'host-placement': {'type': 'balanced',
+                                    'parameters': {'attribute': 'HOSTNAME',
+                                                   'minimum': minimum_hosts}}}
+        job_spec = {'group': group['uuid'],
+                    'command': 'sleep 600'}
+        max_jobs_per_host = 3
+        num_jobs = minimum_hosts * max_jobs_per_host
+        uuids, resp = util.submit_jobs(self.cook_url, job_spec, num_jobs, groups=[group])
+        self.assertEqual(201, resp.status_code, resp.content)
+        try:
+            jobs = util.wait_for_jobs(self.cook_url, uuids, 'running')
+            hosts = [j['instances'][0]['hostname'] for j in jobs]
+            host_count = Counter(hosts)
+            self.assertTrue(len(host_count) >= minimum_hosts)
+            self.assertTrue(max(host_count.values()) <= max_jobs_per_host)
+        finally:
+            util.kill_jobs(self.cook_url, [uuids])
+
+    def test_attribute_equals_hostname_constraint(self):
+        slaves = util.get_mesos_state(self.mesos_url)['slaves']
+        max_slave_cpus = max([s['resources']['cpus'] for s in slaves])
+        task_constraint_cpus = util.settings(self.cook_url)['task-constraints']['cpus']
+        # The largest job we can submit that actually fits on a slave
+        max_cpus = min(max_slave_cpus, task_constraint_cpus)
+        # Use the rest of the machine, plus one half CPU so the large job won't fit
+        # The max is required for cases when the task_constraints are larger than the actual hosts
+        canary_cpus = max(0, max_slave_cpus - task_constraint_cpus) + 0.5
+        group = {'uuid': str(uuid.uuid4()),
+                 'host-placement': {'type': 'attribute-equals',
+                                    'parameters': {'attribute': 'HOSTNAME'}}}
+        canary = util.minimal_job(group=group['uuid'],
+                                  priority=100,
+                                  cpus=canary_cpus,
+                                  command='sleep 600')
+        big_job = util.minimal_job(group=group['uuid'],
+                                   priority=1,
+                                   cpus=max_cpus)
+        uuids, resp = util.submit_jobs(self.cook_url, [canary, big_job], groups=[group])
```

Comment:
It would be nice to have a comment explaining that the `big_job` should never get scheduled because it has a lower priority than the `canary`

Reply:
Done